{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "90c32b4b-abb4-431a-90c2-4ab0bcb71d3d",
   "metadata": {},
   "source": [
    "<html>\n",
    "    <summary></summary>\n",
    "         <div> <p></p> </div>\n",
    "         <div style=\"font-size: 20px; width: 800px;\"> \n",
    "              <h1>\n",
    "               <left>Intro to Bioinformatics in Python: Downloading and Annotating Genomic Sequences.</left>\n",
    "              </h1>\n",
    "              <p><left>============================================================================</left> </p>\n",
    "<pre>Course: BIOM/CBE 480A5, Spring 2025\n",
    "Instructor: Brian Munsky\n",
    "Guest instructor: Alex Popinga\n",
    "Contact Info: munsky@colostate.edu\n",
    "</pre>\n",
    "         </div>\n",
    "    </p>\n",
    "\n",
    "</html>\n",
    "\n",
    "<details>\n",
    "  <summary>Copyright info</summary>\n",
    "\n",
    "```\n",
    "Copyright 2024 Brian Munsky\n",
    "\n",
    "Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:\n",
    "\n",
    "1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.\n",
    "\n",
    "2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.\n",
    "\n",
    "3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.\n",
    "\n",
    "THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n",
    "```\n",
    "<details>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66f8f13d",
   "metadata": {},
   "source": [
    "# Phylogenetic Trees\n",
    "\n",
    "## 3.A) Introduction to phylogenetics\n",
    "\n",
    "Phylogenetic trees, or *phylogenies*, are visual representations of evolutionary relationships.  Phylogenetic reconstruction can also provide additional information, such as the time at which events occurred, the population size through time, and mutation and migration rates.  \n",
    "\n",
    "The data that you are comparing - in our case, this will be sequence data - represent the tips of a phylogenetic tree.  The tips are also referred to as the leaves, taxa, or extant nodes.  \n",
    "The internal nodes, at which branching events occur, represent ancestors of multiple branches, or *lineages*.  \n",
    "  \n",
    "<img src=\"Figures/timetree.png\" alt=\"alt text\" width=\"49%\" style=\"margin-right: 10px;\"><img src=\"Figures/timetree_seq.png\" alt=\"alt text\" width=\"49%\">\n",
    "\n",
    "The branches of phylogenetic trees have different interpretations depending on the type of phylogeny you construct.  Two distinctly different types of phylogenetic trees are:  *model-based* trees and *distance-based* trees.  Two other important distinctions are:  *rooted* and *unrooted* trees.\n",
    "\n",
    "A *rooted* tree has a starting point, or *root*.  An *unrooted* tree does not tell you which split, or branching event, in a tree occurred before any other.  It only shows which tips are more closely related to each other than which other tips.\n",
    "\n",
    "<img src=\"Figures/timetree.png\" alt=\"alt text\" width=\"50%\" style=\"margin-right: 100px;\"><img src=\"Figures/disttree.png\" alt=\"alt text\" width=\"30%\">\n",
    "\n",
    "*Model-based trees* are constructed by modelling the process of evolution through time.  They allow us to do much more complex analyses and, in my opinion, more intuitively represent evolutionary history. \n",
    " \n",
    "However, *distance trees* are built simply based on a distance metric (e.g., number of mutations) between each tip and are therefore much easier to understand and more efficient to compute, so we will begin with them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa1652e",
   "metadata": {},
   "source": [
    "## 3.B) Setting up our analyses \n",
    "\n",
    "First, let's make sure we've imported some important packages that we'll need to run our analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "33813934",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %pip install biopython\n",
    "# %pip install Bio\n",
    "# %pip install git+https://github.com/neherlab/treetime.git\n",
    "# conda install -c bioconda -c conda-forge treetime\n",
    "\n",
    "import Bio\n",
    "from Bio import Entrez, SeqIO, pairwise2, AlignIO, Phylo\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import ProtParam\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.image as mpimg\n",
    "import webbrowser\n",
    "import os\n",
    "import threading\n",
    "\n",
    "from Bio import Align\n",
    "from Bio.Align import Alignment\n",
    "from pymsaviz import MsaViz\n",
    "\n",
    "# Install TimeTree\n",
    "# !pip install timetree\n",
    "# !pip install treetime\n",
    "# !pip install PyQt6\n",
    "# !pip install phylo-treetime\n",
    "import treetime, timetree\n",
    "from treetime import TreeAnc\n",
    "from treetime import TreeTime\n",
    "\n",
    "from __future__ import print_function, division\n",
    "from treetime.utils import parse_dates\n",
    "import treetime.config as ttconf\n",
    "from treetime.treetime import plot_vs_years\n",
    "\n",
    "import csv\n",
    "from datetime import datetime\n",
    "\n",
    "from matplotlib import cm\n",
    "import numpy as np\n",
    "\n",
    "from IPython.core.display import display, HTML"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f64a1777",
   "metadata": {},
   "source": [
    "Ensure our open-reading frame functions are defined."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d6411631",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Function to find ORFs in all frames of a sequence\n",
    "def find_all_orfs(sequence):\n",
    "    orf_list = []           # List of all the ORF we find\n",
    "    for frame in range(3):  # Check each of the three reading frames\n",
    "        orfs = []           # List of ORFs for the current frame   \n",
    "        lastStop = frame    # Position of last stop codon.\n",
    "        for orf in sequence[frame:].translate(to_stop=False).split('*'):     # Translate the curent frame sequence into AA and break up by stop codons\n",
    "            if len(orf) > 30:                                                # Consider ORFs longer than 50 amino acids\n",
    "                start = lastStop + orf.find('M') * 3 if 'M' in orf else None # Update start codon position to next M, \n",
    "                stop = lastStop + (len(orf))*3 if 'M' in orf else None       # Update stop codon to next '*'\n",
    "                if start is not None and stop is not None:                   # In the coding region is >30AA\n",
    "                    orfs.append((start, stop))                                      # Append the start and stop positions for that ORF\n",
    "            lastStop += len(orf)*3 + 3           # Update the position of the last stop codon\n",
    "        orf_list.append(orfs)                    # Append the list of ORFs for the current frame\n",
    "    return orf_list\n",
    "\n",
    "def pickNLargestORF(orfs,N=5):\n",
    "    Lengths = []\n",
    "    Bounds = []\n",
    "    for frame, orf in enumerate(orfs):\n",
    "        for start, stop in orf:\n",
    "            Lengths.append(stop-start)\n",
    "            Bounds.append([start,stop])\n",
    "    enumerated_list = list(enumerate(Lengths))\n",
    "    sortedLengths = sorted(enumerated_list, key=lambda x: abs(x[1]), reverse=True)\n",
    "    Bounds = [Bounds[index] for index, _ in sortedLengths[:N]]\n",
    "    return Bounds\n",
    "\n",
    "# Next, let's find the amino acid sequence for the largest open reading frame from each vaccine\n",
    "def pickLargestORF(orfs):\n",
    "    maxLength = 0;\n",
    "    for frame, orfs in enumerate(orfs):\n",
    "        for start, stop in orfs:\n",
    "            if (stop-start)>maxLength:\n",
    "                largestStart = start\n",
    "                largestStop = stop\n",
    "                maxLength = stop-start\n",
    "            \n",
    "    return largestStart, largestStop"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b228c2e",
   "metadata": {},
   "source": [
    "Here are the sequence data for COVID-19 vaccines from BioNTech/Phizer and Moderna."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6d378e52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# (ALIGNMENT) Lets try to align the virus sequence with the BioNTech/Phizer and Moderna vaccines\n",
    "#      First, lets create the sequences for the two vaccines from the following paper: \n",
    "#      https://github.com/NAalytics/Assemblies-of-putative-SARS-CoV2-spike-encoding-mRNA-sequences-for-\n",
    "#          vaccines-BNT-162b2-and-mRNA-1273/blob/main/Assemblies%20of%20putative%20SARS-CoV2-spike-encoding\n",
    "#          %20mRNA%20sequences%20for%20vaccines%20BNT-162b2%20and%20mRNA-1273.docx.pdf\n",
    "BNT_Phizer = '''GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCGCCACCATGTTCGTGTTCCTGGTGCTGCTGCCTCTGGTGTCCA\n",
    "GCCAGTGTGTGAACCTGACCACCAGAACACAGCTGCCTCCAGCCTACACCAACAGCTTTACCAGAGGCGTGTACTACCCCGACAAGGTGTT\n",
    "CAGATCCAGCGTGCTGCACTCTACCCAGGACCTGTTCCTGCCTTTCTTCAGCAACGTGACCTGGTTCCACGCCATCCACGTGTCCGGCACC\n",
    "AATGGCACCAAGAGATTCGACAACCCCGTGCTGCCCTTCAACGACGGGGTGTACTTTGCCAGCACCGAGAAGTCCAACATCATCAGAGGCT\n",
    "GGATCTTCGGCACCACACTGGACAGCAAGACCCAGAGCCTGCTGATCGTGAACAACGCCACCAACGTGGTCATCAAAGTGTGCGAGTTCCA\n",
    "GTTCTGCAACGACCCCTTCCTGGGCGTCTACTACCACAAGAACAACAAGAGCTGGATGGAAAGCGAGTTCCGGGTGTACAGCAGCGCCAAC\n",
    "AACTGCACCTTCGAGTACGTGTCCCAGCCTTTCCTGATGGACCTGGAAGGCAAGCAGGGCAACTTCAAGAACCTGCGCGAGTTCGTGTTTA\n",
    "AGAACATCGACGGCTACTTCAAGATCTACAGCAAGCACACCCCTATCAACCTCGTGCGGGATCTGCCTCAGGGCTTCTCTGCTCTGGAACC\n",
    "CCTGGTGGATCTGCCCATCGGCATCAACATCACCCGGTTTCAGACACTGCTGGCCCTGCACAGAAGCTACCTGACACCTGGCGATAGCAGC\n",
    "AGCGGATGGACAGCTGGTGCCGCCGCTTACTATGTGGGCTACCTGCAGCCTAGAACCTTCCTGCTGAAGTACAACGAGAACGGCACCATCA\n",
    "CCGACGCCGTGGATTGTGCTCTGGATCCTCTGAGCGAGACAAAGTGCACCCTGAAGTCCTTCACCGTGGAAAAGGGCATCTACCAGACCAG\n",
    "CAACTTCCGGGTGCAGCCCACCGAATCCATCGTGCGGTTCCCCAATATCACCAATCTGTGCCCCTTCGGCGAGGTGTTCAATGCCACCAGA\n",
    "TTCGCCTCTGTGTACGCCTGGAACCGGAAGCGGATCAGCAATTGCGTGGCCGACTACTCCGTGCTGTACAACTCCGCCAGCTTCAGCACCT\n",
    "TCAAGTGCTACGGCGTGTCCCCTACCAAGCTGAACGACCTGTGCTTCACAAACGTGTACGCCGACAGCTTCGTGATCCGGGGAGATGAAGT\n",
    "GCGGCAGATTGCCCCTGGACAGACAGGCAAGATCGCCGACTACAACTACAAGCTGCCCGACGACTTCACCGGCTGTGTGATTGCCTGGAAC\n",
    "AGCAACAACCTGGACTCCAAAGTCGGCGGCAACTACAATTACCTGTACCGGCTGTTCCGGAAGTCCAATCTGAAGCCCTTCGAGCGGGACA\n",
    "TCTCCACCGAGATCTATCAGGCCGGCAGCACCCCTTGTAACGGCGTGGAAGGCTTCAACTGCTACTTCCCACTGCAGTCCTACGGCTTTCA\n",
    "GCCCACAAATGGCGTGGGCTATCAGCCCTACAGAGTGGTGGTGCTGAGCTTCGAACTGCTGCATGCCCCTGCCACAGTGTGCGGCCCTAAG\n",
    "AAAAGCACCAATCTCGTGAAGAACAAATGCGTGAACTTCAACTTCAACGGCCTGACCGGCACCGGCGTGCTGACAGAGAGCAACAAGAAGT\n",
    "TCCTGCCATTCCAGCAGTTTGGCCGGGATATCGCCGATACCACAGACGCCGTTAGAGATCCCCAGACACTGGAAATCCTGGACATCACCCC\n",
    "TTGCAGCTTCGGCGGAGTGTCTGTGATCACCCCTGGCACCAACACCAGCAATCAGGTGGCAGTGCTGTACCAGGACGTGAACTGTACCGAA\n",
    "GTGCCCGTGGCCATTCACGCCGATCAGCTGACACCTACATGGCGGGTGTACTCCACCGGCAGCAATGTGTTTCAGACCAGAGCCGGCTGTC\n",
    "TGATCGGAGCCGAGCACGTGAACAATAGCTACGAGTGCGACATCCCCATCGGCGCTGGAATCTGCGCCAGCTACCAGACACAGACAAACAG\n",
    "CCCTCGGAGAGCCAGAAGCGTGGCCAGCCAGAGCATCATTGCCTACACAATGTCTCTGGGCGCCGAGAACAGCGTGGCCTACTCCAACAAC\n",
    "TCTATCGCTATCCCCACCAACTTCACCATCAGCGTGACCACAGAGATCCTGCCTGTGTCCATGACCAAGACCAGCGTGGACTGCACCATGT\n",
    "ACATCTGCGGCGATTCCACCGAGTGCTCCAACCTGCTGCTGCAGTACGGCAGCTTCTGCACCCAGCTGAATAGAGCCCTGACAGGGATCGC\n",
    "CGTGGAACAGGACAAGAACACCCAAGAGGTGTTCGCCCAAGTGAAGCAGATCTACAAGACCCCTCCTATCAAGGACTTCGGCGGCTTCAAT\n",
    "TTCAGCCAGATTCTGCCCGATCCTAGCAAGCCCAGCAAGCGGAGCTTCATCGAGGACCTGCTGTTCAACAAAGTGACACTGGCCGACGCCG\n",
    "GCTTCATCAAGCAGTATGGCGATTGTCTGGGCGACATTGCCGCCAGGGATCTGATTTGCGCCCAGAAGTTTAACGGACTGACAGTGCTGCC\n",
    "TCCTCTGCTGACCGATGAGATGATCGCCCAGTACACATCTGCCCTGCTGGCCGGCACAATCACAAGCGGCTGGACATTTGGAGCAGGCGCC\n",
    "GCTCTGCAGATCCCCTTTGCTATGCAGATGGCCTACCGGTTCAACGGCATCGGAGTGACCCAGAATGTGCTGTACGAGAACCAGAAGCTGA\n",
    "TCGCCAACCAGTTCAACAGCGCCATCGGCAAGATCCAGGACAGCCTGAGCAGCACAGCAAGCGCCCTGGGAAAGCTGCAGGACGTGGTCAA\n",
    "CCAGAATGCCCAGGCACTGAACACCCTGGTCAAGCAGCTGTCCTCCAACTTCGGCGCCATCAGCTCTGTGCTGAACGATATCCTGAGCAGA\n",
    "CTGGACCCTCCTGAGGCCGAGGTGCAGATCGACAGACTGATCACAGGCAGACTGCAGAGCCTCCAGACATACGTGACCCAGCAGCTGATCA\n",
    "GAGCCGCCGAGATTAGAGCCTCTGCCAATCTGGCCGCCACCAAGATGTCTGAGTGTGTGCTGGGCCAGAGCAAGAGAGTGGACTTTTGCGG\n",
    "CAAGGGCTACCACCTGATGAGCTTCCCTCAGTCTGCCCCTCACGGCGTGGTGTTTCTGCACGTGACATATGTGCCCGCTCAAGAGAAGAAT\n",
    "TTCACCACCGCTCCAGCCATCTGCCACGACGGCAAAGCCCACTTTCCTAGAGAAGGCGTGTTCGTGTCCAACGGCACCCATTGGTTCGTGA\n",
    "CACAGCGGAACTTCTACGAGCCCCAGATCATCACCACCGACAACACCTTCGTGTCTGGCAACTGCGACGTCGTGATCGGCATTGTGAACAA\n",
    "TACCGTGTACGACCCTCTGCAGCCCGAGCTGGACAGCTTCAAAGAGGAACTGGACAAGTACTTTAAGAACCACACAAGCCCCGACGTGGAC\n",
    "CTGGGCGATATCAGCGGAATCAATGCCAGCGTCGTGAACATCCAGAAAGAGATCGACCGGCTGAACGAGGTGGCCAAGAATCTGAACGAGA\n",
    "GCCTGATCGACCTGCAAGAACTGGGGAAGTACGAGCAGTACATCAAGTGGCCCTGGTACATCTGGCTGGGCTTTATCGCCGGACTGATTGC\n",
    "CATCGTGATGGTCACAATCATGCTGTGTTGCATGACCAGCTGCTGTAGCTGCCTGAAGGGCTGTTGTAGCTGTGGCAGCTGCTGCAAGTTC\n",
    "GACGAGGACGATTCTGAGCCCGTGCTGAAGGGCGTGAAACTGCACTACACATGATGACTCGAGCTGGTACTGCATGCACGCAATGCTAGCT\n",
    "GCCCCTTTCCCGTCCTGGGTACCCCGAGTCTCCCCCGACCTCGGGTCCCAGGTATGCTCCCACCTCCACCTGCCCCACTCACCACCTCTGC\n",
    "TAGTTCCAGACACCTCCCAAGCACGCAGCAATGCAGCTCAAAACGCTTAGCCTAGCCACACCCCCACGGGAAACAGCAGTGATTAACCTTT\n",
    "AGCAATAAACGAAAGTTTAACTAAGCTATACTAACCCCAGGGTTGGTCAATTTCGTGCCAGCCACACCCTGGAGCTAGCA'''\n",
    "BNT_Phizer = BNT_Phizer.replace(\"\\n\", \"\")\n",
    "\n",
    "Moderna = '''GGGAAATAAGAGAGAAAAGAAGAGTAAGAAGAAATATAAGACCCCGGCGCCGCCACCATGTTCGTGTTCCTGGTGCTGCTGCCCCTGGTGA\n",
    "GCAGCCAGTGCGTGAACCTGACCACCCGGACCCAGCTGCCACCAGCCTACACCAACAGCTTCACCCGGGGCGTCTACTACCCCGACAAGGT\n",
    "GTTCCGGAGCAGCGTCCTGCACAGCACCCAGGACCTGTTCCTGCCCTTCTTCAGCAACGTGACCTGGTTCCACGCCATCCACGTGAGCGGC\n",
    "ACCAACGGCACCAAGCGGTTCGACAACCCCGTGCTGCCCTTCAACGACGGCGTGTACTTCGCCAGCACCGAGAAGAGCAACATCATCCGGG\n",
    "GCTGGATCTTCGGCACCACCCTGGACAGCAAGACCCAGAGCCTGCTGATCGTGAATAACGCCACCAACGTGGTGATCAAGGTGTGCGAGTT\n",
    "CCAGTTCTGCAACGACCCCTTCCTGGGCGTGTACTACCACAAGAACAACAAGAGCTGGATGGAGAGCGAGTTCCGGGTGTACAGCAGCGCC\n",
    "AACAACTGCACCTTCGAGTACGTGAGCCAGCCCTTCCTGATGGACCTGGAGGGCAAGCAGGGCAACTTCAAGAACCTGCGGGAGTTCGTGT\n",
    "TCAAGAACATCGACGGCTACTTCAAGATCTACAGCAAGCACACCCCAATCAACCTGGTGCGGGATCTGCCCCAGGGCTTCTCAGCCCTGGA\n",
    "GCCCCTGGTGGACCTGCCCATCGGCATCAACATCACCCGGTTCCAGACCCTGCTGGCCCTGCACCGGAGCTACCTGACCCCAGGCGACAGC\n",
    "AGCAGCGGGTGGACAGCAGGCGCGGCTGCTTACTACGTGGGCTACCTGCAGCCCCGGACCTTCCTGCTGAAGTACAACGAGAACGGCACCA\n",
    "TCACCGACGCCGTGGACTGCGCCCTGGACCCTCTGAGCGAGACCAAGTGCACCCTGAAGAGCTTCACCGTGGAGAAGGGCATCTACCAGAC\n",
    "CAGCAACTTCCGGGTGCAGCCCACCGAGAGCATCGTGCGGTTCCCCAACATCACCAACCTGTGCCCCTTCGGCGAGGTGTTCAACGCCACC\n",
    "CGGTTCGCCAGCGTGTACGCCTGGAACCGGAAGCGGATCAGCAACTGCGTGGCCGACTACAGCGTGCTGTACAACAGCGCCAGCTTCAGCA\n",
    "CCTTCAAGTGCTACGGCGTGAGCCCCACCAAGCTGAACGACCTGTGCTTCACCAACGTGTACGCCGACAGCTTCGTGATCCGTGGCGACGA\n",
    "GGTGCGGCAGATCGCACCCGGCCAGACAGGCAAGATCGCCGACTACAACTACAAGCTGCCCGACGACTTCACCGGCTGCGTGATCGCCTGG\n",
    "AACAGCAACAACCTCGACAGCAAGGTGGGCGGCAACTACAACTACCTGTACCGGCTGTTCCGGAAGAGCAACCTGAAGCCCTTCGAGCGGG\n",
    "ACATCAGCACCGAGATCTACCAAGCCGGCTCCACCCCTTGCAACGGCGTGGAGGGCTTCAACTGCTACTTCCCTCTGCAGAGCTACGGCTT\n",
    "CCAGCCCACCAACGGCGTGGGCTACCAGCCCTACCGGGTGGTGGTGCTGAGCTTCGAGCTGCTGCACGCCCCAGCCACCGTGTGTGGCCCC\n",
    "AAGAAGAGCACCAACCTGGTGAAGAACAAGTGCGTGAACTTCAACTTCAACGGCCTTACCGGCACCGGCGTGCTGACCGAGAGCAACAAGA\n",
    "AATTCCTGCCCTTTCAGCAGTTCGGCCGGGACATCGCCGACACCACCGACGCTGTGCGGGATCCCCAGACCCTGGAGATCCTGGACATCAC\n",
    "CCCTTGCAGCTTCGGCGGCGTGAGCGTGATCACCCCAGGCACCAACACCAGCAACCAGGTGGCCGTGCTGTACCAGGACGTGAACTGCACC\n",
    "GAGGTGCCCGTGGCCATCCACGCCGACCAGCTGACACCCACCTGGCGGGTCTACAGCACCGGCAGCAACGTGTTCCAGACCCGGGCCGGTT\n",
    "GCCTGATCGGCGCCGAGCACGTGAACAACAGCTACGAGTGCGACATCCCCATCGGCGCCGGCATCTGTGCCAGCTACCAGACCCAGACCAA\n",
    "TTCACCCCGGAGGGCAAGGAGCGTGGCCAGCCAGAGCATCATCGCCTACACCATGAGCCTGGGCGCCGAGAACAGCGTGGCCTACAGCAAC\n",
    "AACAGCATCGCCATCCCCACCAACTTCACCATCAGCGTGACCACCGAGATTCTGCCCGTGAGCATGACCAAGACCAGCGTGGACTGCACCA\n",
    "TGTACATCTGCGGCGACAGCACCGAGTGCAGCAACCTGCTGCTGCAGTACGGCAGCTTCTGCACCCAGCTGAACCGGGCCCTGACCGGCAT\n",
    "CGCCGTGGAGCAGGACAAGAACACCCAGGAGGTGTTCGCCCAGGTGAAGCAGATCTACAAGACCCCTCCCATCAAGGACTTCGGCGGCTTC\n",
    "AACTTCAGCCAGATCCTGCCCGACCCCAGCAAGCCCAGCAAGCGGAGCTTCATCGAGGACCTGCTGTTCAACAAGGTGACCCTAGCCGACG\n",
    "CCGGCTTCATCAAGCAGTACGGCGACTGCCTCGGCGACATAGCCGCCCGGGACCTGATCTGCGCCCAGAAGTTCAACGGCCTGACCGTGCT\n",
    "GCCTCCCCTGCTGACCGACGAGATGATCGCCCAGTACACCAGCGCCCTGTTAGCCGGAACCATCACCAGCGGCTGGACTTTCGGCGCTGGA\n",
    "GCCGCTCTGCAGATCCCCTTCGCCATGCAGATGGCCTACCGGTTCAACGGCATCGGCGTGACCCAGAACGTGCTGTACGAGAACCAGAAGC\n",
    "TGATCGCCAACCAGTTCAACAGCGCCATCGGCAAGATCCAGGACAGCCTGAGCAGCACCGCTAGCGCCCTGGGCAAGCTGCAGGACGTGGT\n",
    "GAACCAGAACGCCCAGGCCCTGAACACCCTGGTGAAGCAGCTGAGCAGCAACTTCGGCGCCATCAGCAGCGTGCTGAACGACATCCTGAGC\n",
    "CGGCTGGACCCTCCCGAGGCCGAGGTGCAGATCGACCGGCTGATCACTGGCCGGCTGCAGAGCCTGCAGACCTACGTGACCCAGCAGCTGA\n",
    "TCCGGGCCGCCGAGATTCGGGCCAGCGCCAACCTGGCCGCCACCAAGATGAGCGAGTGCGTGCTGGGCCAGAGCAAGCGGGTGGACTTCTG\n",
    "CGGCAAGGGCTACCACCTGATGAGCTTTCCCCAGAGCGCACCCCACGGAGTGGTGTTCCTGCACGTGACCTACGTGCCCGCCCAGGAGAAG\n",
    "AACTTCACCACCGCCCCAGCCATCTGCCACGACGGCAAGGCCCACTTTCCCCGGGAGGGCGTGTTCGTGAGCAACGGCACCCACTGGTTCG\n",
    "TGACCCAGCGGAACTTCTACGAGCCCCAGATCATCACCACCGACAACACCTTCGTGAGCGGCAACTGCGACGTGGTGATCGGCATCGTGAA\n",
    "CAACACCGTGTACGATCCCCTGCAGCCCGAGCTGGACAGCTTCAAGGAGGAGCTGGACAAGTACTTCAAGAATCACACCAGCCCCGACGTG\n",
    "GACCTGGGCGACATCAGCGGCATCAACGCCAGCGTGGTGAACATCCAGAAGGAGATCGATCGGCTGAACGAGGTGGCCAAGAACCTGAACG\n",
    "AGAGCCTGATCGACCTGCAGGAGCTGGGCAAGTACGAGCAGTACATCAAGTGGCCCTGGTACATCTGGCTGGGCTTCATCGCCGGCCTGAT\n",
    "CGCCATCGTGATGGTGACCATCATGCTGTGCTGCATGACCAGCTGCTGCAGCTGCCTGAAGGGCTGTTGCAGCTGCGGCAGCTGCTGCAAG\n",
    "TTCGACGAGGACGACAGCGAGCCCGTGCTGAAGGGCGTGAAGCTGCACTACACCTGATAATAGGCTGGAGCCTCGGTGGCCTAGCTTCTTG\n",
    "CCCCTTGGGCCTCCCCCCAGCCCCTCCTCCCCTTCCTGCACCCGTACCCCCGTGGTCTTTGAATAAAGTCTGAGTGGGCGGCAAAAAAAAA'''\n",
    "\n",
    "Moderna = Moderna.replace(\"\\n\", \"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dda9312e",
   "metadata": {},
   "source": [
    "Now, let's retrieve our data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a750f91b-5d63-4be6-a627-2864fd3d56e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#(PRELIMINARIES) Let's download the COVID19 virus sequence from NCBI\n",
    "# Provide your email for accessing NCBI\n",
    "Entrez.email = \"brian.munsky@colostate.edu\"  # Insert your email here\n",
    "\n",
    "def get_genbank(accession_number):\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"gb\", retmode=\"text\")\n",
    "    record = SeqIO.read(handle, \"genbank\")\n",
    "    handle.close()\n",
    "    return record\n",
    "\n",
    "corona_virus = get_genbank(\"MN908947\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b271519d-c368-4570-a437-2f002f11d8f3",
   "metadata": {},
   "source": [
    "## 3.C) Distance-based phylogenetic trees\n",
    "\n",
    "### UPGMA\n",
    "We're going to start with a distance method called *Unweighted Pair Group Method with Arithmetic Mean (UPGMA)*.  \n",
    "\n",
    "$d(A,C) \\leq max(d(A,B),d(B,C))$\n",
    "\n",
    "Features of UPGMA:\n",
    "- Produces an *ultrametric tree*, where all tips are considered to have equal distance to the root (if there is a root)\n",
    "- Assumes constant rate of evolution across all lineages\n",
    "- Iteratively merges pairs of taxa and updates distance matrix\n",
    "- Simple and fast, can be unrealistic\n",
    "\n",
    "Let's define our UPGMA function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e8f82b46-b673-4697-a858-58e9e84080dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeTree(alignedFile):\n",
    "    # read our alignment file.\n",
    "    align = AlignIO.read(alignedFile, 'fasta')\n",
    "\n",
    "    # define sequence distance metric\n",
    "    calculator = DistanceCalculator('identity')\n",
    "    dm  = calculator.get_distance(align)\n",
    "\n",
    "    # construct tree\n",
    "    constructor = DistanceTreeConstructor()\n",
    "    return constructor.upgma(dm), dm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c609a57a",
   "metadata": {},
   "source": [
    "Let's make an ultrametric tree for our aligned BLAST results from Module 2C."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a273669a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We will use the alignment file that we found in the last lesson:\n",
    "alignedFile = \"blast_file_aligned.afa\"\n",
    "\n",
    "tree, dm = makeTree(alignedFile)\n",
    "\n",
    "# plot tree\n",
    "fig, ax = plt.subplots(figsize=(8, 30))  # Adjust size as needed\n",
    "Phylo.draw(tree,axes=ax)\n",
    "Phylo.draw_ascii(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38b67e03",
   "metadata": {},
   "source": [
    "Let's take a look at the distance matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b7fb287-6066-4752-9232-f6e4b3ccd08a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The numerical distances can be found in the distance matrix \"dm\"\n",
    "print(dm)\n",
    "\n",
    "print(f'The phylogenetic distance between {dm.names[0]} and {dm.names[4]} is {dm[0,4]}')\n",
    "print(f'The phylogenetic ditance between {dm.names[1]} and {dm.names[4]} is {dm[1,4]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "838cb70c-3c14-49d7-9072-060f66842b15",
   "metadata": {},
   "source": [
    "### Explore different covid variants\n",
    "\n",
    "Let's focus our analysis a little more closely onto some famous variants that were in the news during the early stages of the COVID19 pandemic:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aaeefcd5-822e-4021-9125-2e7cd357d784",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create a dictionary to map the sequence IDs to the sequence names\n",
    "variantDict = {\"wuhan\":'MN908947',\n",
    "            \"delta\":'OK091006',\n",
    "            \"beta\":'MZ433432',\n",
    "            \"alpha\":'MZ773928',\n",
    "            \"omicron\":'PP028445',\n",
    "            \"OP208953\":'OP208953',\n",
    "            \"MT628218\":'MT628218'}\n",
    "\n",
    "sequenceEntry = {}\n",
    "for variant in variantDict:\n",
    "    print(f\"Downloading {variant} variant using accession number {variantDict[variant]}\")\n",
    "    sequenceEntry[variant] = get_genbank(variantDict[variant])\n",
    "    # sequenceEntry.append(get_genbank(variantDict[variant]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eee4b34",
   "metadata": {},
   "source": [
    "Let's also add the original Moderna vaccine to the mix so we can see what sequence it was most similar to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c8bb744",
   "metadata": {},
   "outputs": [],
   "source": [
    "Moderna_orf_positions = find_all_orfs(Seq(Moderna))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b7bad1c",
   "metadata": {},
   "source": [
    "Next, we will use a loop to write the protein sequences of all the different variants into a single FASTA file.  Note how we are including both the name and accession number varantDict[variant] and the translated sequence in each entry. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcfadb6e-ff4e-4fc2-bf10-fc5db6ff9374",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variant_file_fasta = \"covid_variants.fa\"\n",
    "with open(variant_file_fasta, \"w\") as handle:\n",
    "    for variant in variantDict:\n",
    "        # Find and record the third largest ORF for each sequence\n",
    "        orf_positions = find_all_orfs(sequenceEntry[variant].seq)\n",
    "        longestORFbounds = pickNLargestORF(orf_positions,5)\n",
    "        proteinOfInterest = sequenceEntry[variant].seq[longestORFbounds[2][0]:].translate(to_stop=True)\n",
    "        handle.write(f\">{variant}:{variantDict[variant]}\\n{proteinOfInterest}\\n\")\n",
    "    startMod, stopMod = pickLargestORF(Moderna_orf_positions)\n",
    "    proteinOfInterest = Seq(Moderna[startMod:stopMod]).translate(to_stop=True)\n",
    "    handle.write(f\">Moderna\\n{proteinOfInterest}\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8d5f3c1",
   "metadata": {},
   "source": [
    "Now, we'll align our sequence data using MUSCLE, as we've done before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1445a8b9-39f9-42a3-8b74-f30b9c28e57e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Now that we have a fasta file, we can run MUSCLE again to \n",
    "# compute and plot its alignment:\n",
    "alignedFile = \"covid_variants_aligned.afa\"\n",
    "\n",
    "# Use a terminal command (bash) to run MUSCLE in to align the sequences.        \n",
    "bashCommand = \"muscle -align \" + variant_file_fasta + \" -output \" + alignedFile\n",
    "os.system(bashCommand)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a44cddc3",
   "metadata": {},
   "source": [
    "It's always nice to visualise the alignment to give us a quick idea of sequence similarity (and ensure we've grabbed the correct data and used the correct alignment method for our data), so let's use MsaViz again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4936e8b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make and save an alignment figure\n",
    "mv = MsaViz(alignedFile, wrap_length=150, show_count=True)\n",
    "mv.savefig(\"Variant_Alignment.png\")\n",
    "\n",
    "# Load the image\n",
    "img = mpimg.imread('Variant_Alignment.png')\n",
    "\n",
    "# Display the image\n",
    "fig = plt.figure(figsize=(80, 80), dpi=200)\n",
    "plt.imshow(img)\n",
    "plt.axis('off')  # Hide axis\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48d5fd87",
   "metadata": {},
   "source": [
    "We'll make a distance tree of our covid variants using UPGMA..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "578d9f8b-9ba6-4d4e-b325-d4f1c095fc20",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Compute the phylogenetic tree using our previous function:\n",
    "tree = makeTree(alignedFile)\n",
    "# plot tree\n",
    "Phylo.draw_ascii(tree[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54970c84",
   "metadata": {},
   "source": [
    "...and look at the distance matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "450cbec7-52a1-4b5d-9b6d-97d3bd3ec712",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The numerical distances can be found in the distance matrix \"dm\"\n",
    "print(tree[1])\n",
    "\n",
    "print(f'The phylogenetic distance between {tree[1].names[0]} and {tree[1].names[4]} is {tree[1][0,4]}')\n",
    "print(f'The phylogenetic distance between {tree[1].names[1]} and {tree[1].names[4]} is {tree[1][1,4]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8eff60fc",
   "metadata": {},
   "source": [
    "### Tree files and notation\n",
    "\n",
    "Let's save our tree to a file.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6787c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the tree in Newick format\n",
    "Phylo.write(tree[0], \"covid_variants_aligned.newick\", \"newick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bf20e41",
   "metadata": {},
   "source": [
    "Newick and nexus are common tree file formats.  Both store the tree using simple paranthentical notation.  The innermost parantheses show the closest relationships between taxa, and numbers after colons show the branch lengths (distances, for distance trees, and duration of time for time trees.  For example, a 4 taxa tree with tips A, B, C, and D, might be saved in a newick file as **(A:0.1,(B:0.2,(C:0.3,D:0.4):0.5):0.1);** and look like:\n",
    "\n",
    "<img src=\"Figures/ABCD.png\" alt=\"alt text\" width=\"30%\">\n",
    "\n",
    "#### **Challenge Questions!** \n",
    "\n",
    "1. Can you draw the tree for (A,B,(C,D))?  \n",
    "\n",
    "2. Can you write the newick format for the tree topology shown in the image below (ignore branch lengths)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d2b30f41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    Select an Image:\n",
       "    <input type=\"radio\" name=\"images\" onclick=\"show1t();\" checked>Question 2 tree</input>\n",
       "    <input type=\"radio\" name=\"images\" onclick=\"show2t();\">Answer 1</input>\n",
       "    <input type=\"radio\" name=\"images\" onclick=\"show3t();\">Answer 2</input>\n",
       "\n",
       "\n",
       "    <div id=\"image1t\">\n",
       "      <img src=\"Figures/covid_tree.png\" width=\"30%\">\n",
       "    </div>\n",
       "    <div id=\"image2t\">\n",
       "      <img src=\"Figures/polytomy.png\" width=\"30%\">\n",
       "    </div>\n",
       "    <div id=\"image3t\">\n",
       "      ((Moderna,(beta,((delta,(MT628218,wuhan)),alpha))),(OP208953,omicron));\n",
       "    </div>\n",
       "    \n",
       "    <script>\n",
       "    function show1t(){\n",
       "      document.getElementById('image1t').style.display ='block';\n",
       "      document.getElementById('image2t').style.display ='none';\n",
       "      document.getElementById('image3t').style.display ='none';\n",
       "    }\n",
       "    function show2t(){\n",
       "      document.getElementById('image2t').style.display ='block';\n",
       "      document.getElementById('image1t').style.display ='none';\n",
       "      document.getElementById('image3t').style.display ='none';\n",
       "    }\n",
       "    function show3t(){\n",
       "      document.getElementById('image3t').style.display ='block';\n",
       "      document.getElementById('image1t').style.display ='none';\n",
       "      document.getElementById('image2t').style.display ='none';\n",
       "    }\n",
       "    show1t()\n",
       "    </script>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Toggle answers\n",
    "\n",
    "html_tree = \"\"\"\n",
    "    Select an Image:\n",
    "    <input type=\"radio\" name=\"images\" onclick=\"show1t();\" checked>Question 2 tree</input>\n",
    "    <input type=\"radio\" name=\"images\" onclick=\"show2t();\">Answer 1</input>\n",
    "    <input type=\"radio\" name=\"images\" onclick=\"show3t();\">Answer 2</input>\n",
    "\n",
    "\n",
    "    <div id=\"image1t\">\n",
    "      <img src=\"Figures/covid_tree.png\" width=\"30%\">\n",
    "    </div>\n",
    "    <div id=\"image2t\">\n",
    "      <img src=\"Figures/polytomy.png\" width=\"30%\">\n",
    "    </div>\n",
    "    <div id=\"image3t\">\n",
    "      ((Moderna,(beta,((delta,(MT628218,wuhan)),alpha))),(OP208953,omicron));\n",
    "    </div>\n",
    "    \n",
    "    <script>\n",
    "    function show1t(){\n",
    "      document.getElementById('image1t').style.display ='block';\n",
    "      document.getElementById('image2t').style.display ='none';\n",
    "      document.getElementById('image3t').style.display ='none';\n",
    "    }\n",
    "    function show2t(){\n",
    "      document.getElementById('image2t').style.display ='block';\n",
    "      document.getElementById('image1t').style.display ='none';\n",
    "      document.getElementById('image3t').style.display ='none';\n",
    "    }\n",
    "    function show3t(){\n",
    "      document.getElementById('image3t').style.display ='block';\n",
    "      document.getElementById('image1t').style.display ='none';\n",
    "      document.getElementById('image2t').style.display ='none';\n",
    "    }\n",
    "    show1t()\n",
    "    </script>\n",
    "\"\"\"\n",
    "display(HTML(html_tree))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8894e0fc",
   "metadata": {},
   "source": [
    "### But how have these covid variants changed through time?  *When* did they diverge from each other?  Did the different variants mutate at the same rate?\n",
    "\n",
    "To glean information about evolution through time, we'll need to construct *time trees* based on evolutionary models using *model-based phylogenetics*.  \n",
    "When we use models to make trees, we can utilise additional information, such as the times at which the covid samples were collected.\n",
    "\n",
    "Let's get those dates from GenBank so we can use them to make a time tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f5e7260",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set your Entrez email (required by NCBI)\n",
    "Entrez.email = \"your_email@example.com\"\n",
    "\n",
    "def get_genbank_from_sequence(nucleotide_query):\n",
    "    # Run a BLAST search on NCBI\n",
    "    result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", nucleotide_query)\n",
    "\n",
    "    # Parse BLAST results\n",
    "    blast_record = NCBIXML.read(result_handle)\n",
    "\n",
    "    # Get the best hit (first one in the list)\n",
    "    if not blast_record.alignments:\n",
    "        print(\"No matching GenBank record found.\")\n",
    "        return None\n",
    "\n",
    "    best_hit = blast_record.alignments[0]\n",
    "    accession_number = best_hit.accession\n",
    "\n",
    "    # Fetch the GenBank record using the best hit's accession number\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=accession_number, rettype=\"gb\", retmode=\"text\")\n",
    "    record = SeqIO.read(handle, \"genbank\")\n",
    "    handle.close()\n",
    "\n",
    "    # Extract submission date\n",
    "    submission_date = record.annotations.get(\"date\", \"Unknown date\")\n",
    "\n",
    "    return record, submission_date\n",
    "\n",
    "# BLAST our Moderna sequence and get date\n",
    "query = Moderna\n",
    "genbank_record, query_date = get_genbank_from_sequence(query)\n",
    "\n",
    "if genbank_record:\n",
    "    print(f\"Best match: {genbank_record.id}\")\n",
    "    print(f\"Description: {genbank_record.description}\")\n",
    "    print(f\"Submission Date: {query_date}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30458803",
   "metadata": {},
   "source": [
    "And let's go ahead and add our Moderna sequence and its date to our dictionary of covid variants (variantDict):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "338bbf72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new dictionary entry for the query sequence\n",
    "query_variant_name = \"Moderna\"  # You can modify this name if needed\n",
    "query_accession = genbank_record.id\n",
    "\n",
    "# Ensure variantDict and sequenceEntry contain the query data\n",
    "variantDict[query_variant_name] = query_variant_name  # Adding query name\n",
    "sequenceEntry[query_variant_name] = genbank_record  # Store the GenBank record\n",
    "\n",
    "variantDict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98162f1e",
   "metadata": {},
   "source": [
    "Great!  Now we're ready to do:\n",
    "\n",
    "## 3.D) Model-based Phylogeneties\n",
    "Using models to reconstruct our trees allows us to infer parameters such as mutation rates, changes in population sizes through time, horizontal gene transfer events, rates of viral infection and recovery, times of evolutionary events, and much more!  \n",
    "We can also quantify our *uncertainties* about all of these inferences, including the tree itself and every branching event.\n",
    "\n",
    "<img src=\"Figures/ml.png\" alt=\"alt text\" width=\"42%\" style=\"margin-right: 50px;\"><img src=\"Figures/Bayes.png\" alt=\"alt text\" width=\"52%\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9a198ea",
   "metadata": {},
   "source": [
    "### Maximum likelihood and Bayesian inference\n",
    "We are going to discuss *maximum likelihood* and *Bayesian inference* in more detail later in the course.  Here, I'll briefly describe these two methods in context of phylogenetics.\n",
    "\n",
    "### Maximum likelihood\n",
    "The maximum likelihood approach samples a number of trees and computes the likelihood of the data given each tree,\n",
    "\n",
    "$Pr(data|tree)$.\n",
    "\n",
    "The answer of which tree best represents your data is the tree which maximises the likelihood.  In the above figure (left), the tree **(A,(C,(B,D)))** on the right has the highest likelihood peak.\n",
    "\n",
    "### Bayesian inference\n",
    "Bayesian inference gives you not just one tree as the only explanation for the evolutionary trajectories of your data, but rather, it samples many trees (using techniques such as Markov Chain Monte Carlo) and computes the probabilities that each tree represents the process of evolution for your data.  In other words, Bayesian inference provides you a *posterior probability distribution* of trees such that: \n",
    "\n",
    "$P(tree|data) = \\frac{Pr(data|tree)P(model)}{P(data)}$.\n",
    "\n",
    "As you can see, Bayesian inference incorporates the likelihood term used in the maximum likelihood method. However, it also utilises *prior probabilities* which represent any additional information we may have about our model and data a priori.  This includes tree priors, that describe the branching process through time, and substitution processes such as *relaxed molecular clocks* (allowing for changes in the rates of evolution across branches and across time).\n",
    "\n",
    "Note: The normalising constant $P(data)$ is more often than not marginalised over using an algorithm called Felsenstein's pruning (or peeling) algorithm.\n",
    "\n",
    "The other cool thing about Bayesian inference is that it automatically provides you meaures of uncertainty about your trees, since it works directly with probabilities, as well as any other model parameters of interest like the ones we listed at the start of this section (3.D) Model-based Phylogenetics).  In the above figure (right), you can see some of the sampled trees from a posterior distribution of trees inferred from an alignment of 4 sequences.  The tree on the left **((A,B),(C,D))** has by far the highest posterior probability and thus likely is the best representation of our data.\n",
    "\n",
    "Bayesian phylogenetics may be conceptually simple, but its implementation is more complex. \n",
    "\n",
    "Today, we're going to use *TimeTree*, a maximum likelihood approach.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ffe2520",
   "metadata": {},
   "source": [
    "### **Challenge Questions!**\n",
    "\n",
    "1) How many possible tree topologies (disregard branch lengths) are possible for a 4 taxa tree?\n",
    "\n",
    "2) How many possible topologies are there for a 6 taxa tree?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bf8f0b09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    Select an Image:\n",
       "    <input type=\"radio\" name=\"images\" onclick=\"show1s();\" checked>4 taxa tree</input>\n",
       "    <input type=\"radio\" name=\"images\" onclick=\"show2s();\">Answer 1</input>\n",
       "    <input type=\"radio\" name=\"images\" onclick=\"show3s();\">Answer 2</input>\n",
       "\n",
       "\n",
       "    <div id=\"image1s\">\n",
       "      <img src=\"Figures/4proteins_1tree.png\">\n",
       "    </div>\n",
       "    <div id=\"image2s\">\n",
       "      <img src=\"Figures/4proteins.png\">\n",
       "    </div>\n",
       "    <div id=\"image3s\">\n",
       "      <img src=\"Figures/6proteins.png\">\n",
       "    </div>\n",
       "    \n",
       "    <script>\n",
       "    function show1s(){\n",
       "      document.getElementById('image1s').style.display ='block';\n",
       "      document.getElementById('image2s').style.display ='none';\n",
       "      document.getElementById('image3s').style.display ='none';\n",
       "    }\n",
       "    function show2s(){\n",
       "      document.getElementById('image2s').style.display ='block';\n",
       "      document.getElementById('image1s').style.display ='none';\n",
       "      document.getElementById('image3s').style.display ='none';\n",
       "    }\n",
       "    function show3s(){\n",
       "      document.getElementById('image3s').style.display ='block';\n",
       "      document.getElementById('image1s').style.display ='none';\n",
       "      document.getElementById('image2s').style.display ='none';\n",
       "    }\n",
       "    show1s()\n",
       "    </script>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Toggle answers\n",
    "\n",
    "htmls = \"\"\"\n",
    "    Select an Image:\n",
    "    <input type=\"radio\" name=\"images\" onclick=\"show1s();\" checked>4 taxa tree</input>\n",
    "    <input type=\"radio\" name=\"images\" onclick=\"show2s();\">Answer 1</input>\n",
    "    <input type=\"radio\" name=\"images\" onclick=\"show3s();\">Answer 2</input>\n",
    "\n",
    "\n",
    "    <div id=\"image1s\">\n",
    "      <img src=\"Figures/4proteins_1tree.png\">\n",
    "    </div>\n",
    "    <div id=\"image2s\">\n",
    "      <img src=\"Figures/4proteins.png\">\n",
    "    </div>\n",
    "    <div id=\"image3s\">\n",
    "      <img src=\"Figures/6proteins.png\">\n",
    "    </div>\n",
    "    \n",
    "    <script>\n",
    "    function show1s(){\n",
    "      document.getElementById('image1s').style.display ='block';\n",
    "      document.getElementById('image2s').style.display ='none';\n",
    "      document.getElementById('image3s').style.display ='none';\n",
    "    }\n",
    "    function show2s(){\n",
    "      document.getElementById('image2s').style.display ='block';\n",
    "      document.getElementById('image1s').style.display ='none';\n",
    "      document.getElementById('image3s').style.display ='none';\n",
    "    }\n",
    "    function show3s(){\n",
    "      document.getElementById('image3s').style.display ='block';\n",
    "      document.getElementById('image1s').style.display ='none';\n",
    "      document.getElementById('image2s').style.display ='none';\n",
    "    }\n",
    "    show1s()\n",
    "    </script>\n",
    "\"\"\"\n",
    "display(HTML(htmls))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cb200d6",
   "metadata": {},
   "source": [
    "\n",
    "### Time Trees\n",
    "To build more complex models and infer additional information (besides simple distance metrics), we need trees that incorporate a model of evolution through time.  In the case of time trees, each branch length is a measurement of time that has passed (rather than distances).\n",
    "\n",
    "First, let's format our data for TimeTree.  In model-based phylogenetics, if we have time data, we can include it as part of the model of evolution through time - and thus we take this information into account when we construct our trees!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cb334a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the dates of the sequences\n",
    "for variant in variantDict:\n",
    "    print(variant + \",\" + sequenceEntry[variant].annotations['date']) \n",
    "    \n",
    "# Convert the date to decimal year for TimeTree\n",
    "def date_to_decimal(date_str):\n",
    "    \"\"\"Convert a date in DD-MMM-YYYY format to decimal year.\"\"\"\n",
    "    date = datetime.strptime(date_str, \"%d-%b-%Y\")  # Correct format for DD-MMM-YYYY\n",
    "    year_start = datetime(date.year, 1, 1)\n",
    "    year_end = datetime(date.year + 1, 1, 1)\n",
    "    decimal_year = date.year + (date - year_start).total_seconds() / (year_end - year_start).total_seconds()\n",
    "    return round(decimal_year, 6)  # Rounded for better precision\n",
    "\n",
    "print(\"\\n\")\n",
    "    \n",
    "# Open the file in write mode\n",
    "with open(\"metadata_dates.csv\", mode=\"w\", newline=\"\") as file:\n",
    "    writer = csv.writer(file)\n",
    "    \n",
    "    # Write the header\n",
    "    writer.writerow([\"name\", \"date\"])\n",
    "    \n",
    "    # Iterate over the dictionary and write each variant and its date\n",
    "    for variant in variantDict:\n",
    "        date = sequenceEntry[variant].annotations.get('date') \n",
    "        date = date_to_decimal(date)\n",
    "        accessions = ','.join(sequenceEntry[variant].annotations.get('accessions'))\n",
    "        \n",
    "        # Check the names and conversion of dates to decimal years\n",
    "        name_and_date = variant + \":\" + accessions + \",\" + str(date)\n",
    "        print(name_and_date) \n",
    "        \n",
    "        # Write name and date as separate columns for TreeTime\n",
    "        writer.writerow([f\"{variant}:{accessions}\", date])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5179fa13",
   "metadata": {},
   "source": [
    "Now let's import our data into TimeTree and create time trees using maximum likelihood!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca93d995",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_axes(fig, axs):\n",
    "    axs[0].set_axis_off()\n",
    "    axs[1].tick_params(labelsize=14)\n",
    "    axs[1].set_ylabel(\"root-to-tip distance\", fontsize=16)\n",
    "    axs[1].set_xlabel(\"date\", fontsize=16)\n",
    "    fig.tight_layout()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    # Load our alignment of COVID-19 variants\n",
    "    plt.ion()\n",
    "    base_name = 'covid_variants_aligned'\n",
    "\n",
    "    # Parse the dates we converted to decimal years\n",
    "    dates = parse_dates('metadata_dates.csv')\n",
    "    \n",
    "    # Choose the substitution model.  Jukes-Cantor is the most basic and fastes; it assumes \n",
    "    # equal substitution rates between all 4 nucleotides and no rate variation.  Others: HKY, GTR, etc.\n",
    "    tt = TreeTime(gtr='Jukes-Cantor', tree = base_name+'.newick',\n",
    "                  aln = base_name+'.afa', verbose = 1, dates = dates)\n",
    "\n",
    "    # Rerooting can be done along with the tree time inference\n",
    "    tt.run(root=\"best\", branch_length_mode='input', max_iter=2)\n",
    "    # if more complicated models (relaxed clocks, coalescent models) are to be used\n",
    "    # or you want to resolve polytomies, treetime needs to be run for\n",
    "    # several iterations, for example as\n",
    "    #tt.run(root=\"best\", resolve_polytomies=True, max_iter=2)\n",
    "\n",
    "    # Convert the branch length to units in years and redraw\n",
    "    tt.branch_length_to_years()\n",
    "    fig, axs = plt.subplots(1,2, figsize=(12,6))\n",
    "    Phylo.draw(tt.tree, label_func=lambda x:'', show_confidence=True, axes=axs[1])\n",
    "    axs[1].set_title(\"Tree: units are years\", fontsize=18)\n",
    "    axs[0].tick_params(labelsize=14)\n",
    "    axs[1].tick_params(labelsize=14)\n",
    "    fig.tight_layout()\n",
    "\n",
    "    # Plot tree along time axis (in decimal years) and label tips with strain names (exclude internal nodes) \n",
    "    plot_vs_years(tt, label_func=lambda x:x.name if x.is_terminal() else '', show_confidence=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8362e42c",
   "metadata": {},
   "source": [
    "### Evolutionary rate variation \n",
    "\n",
    "Let's try using a relaxed molecular clock model (allows for evolutionary rates to vary across time and lineages)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7913535b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This script illustrates the use of the relaxed clock model in TreeTime.\n",
    "TreeTime currently only supports a normally distributed clock model.\n",
    "\"\"\"\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    # Load our COVID-19 alignment and parse the sample dates\n",
    "    plt.ion()\n",
    "    base_name = 'covid_variants_aligned'\n",
    "    dates = parse_dates('metadata_dates.csv')\n",
    "\n",
    "    # Choose a substitution model and instantiate treetime\n",
    "    tt_relaxed = TreeTime(gtr='Jukes-Cantor', tree = base_name+'.newick',\n",
    "                        aln = base_name+'.afa', verbose = 4, dates = dates)\n",
    "\n",
    "    # This example uses an autocorrelated molecular clock with normal prior and parent-child coupling\n",
    "    # the parameter slack penalizes rate deviations from the average rate\n",
    "    # couplings penalize rate changes between parent and child nodes.\n",
    "    tt_relaxed.run(root='best', relaxed_clock={\"slack\":5.0, \"coupling\":1.0}, max_iter=3,\n",
    "                   resolve_polytomies=True, Tc=0, time_marginal=False)\n",
    "\n",
    "\n",
    "    ##############\n",
    "    # PLOTTING\n",
    "    ##############\n",
    "\n",
    "    # Draw trees inferred with the relaxed model\n",
    "    fig = plt.figure()\n",
    "    ax = plt.subplot(111)\n",
    "    vmin, vmax = 0.5, 1.5 # color branches according to the rate deviation\n",
    "    for n in tt_relaxed.tree.find_clades():\n",
    "        if n.up:\n",
    "            n.color = [int(x*255) for x in cm.cool((min(max(vmin, n.branch_length_interpolator.gamma),vmax)-vmin)/(vmax-vmin))[:3]]\n",
    "        else:\n",
    "            n.color = [200,200,200]\n",
    "\n",
    "    ax.set_title(\"relaxed clock\")\n",
    "    Phylo.draw(tt_relaxed.tree, axes=ax, show_confidence=False, label_func = lambda x:'')\n",
    "\n",
    "    # Scatter branch stretch against the rate multiplier of the branch.\n",
    "    # They are expected to have a positive correlation\n",
    "    # 1) Collect the optimal branch lenght (called mutation_length),\n",
    "    #    the timetree branch length (clock_length) and the inferred rate multiplier gamma\n",
    "    branch_lengths = []\n",
    "    for n in tt_relaxed.tree.find_clades():\n",
    "        if n.up:\n",
    "            branch_lengths.append((n.mutation_length, n.clock_length, n.branch_length_interpolator.gamma))\n",
    "\n",
    "\n",
    "    # 2) Plot the difference between optimal and timetree branch length vs the rate multiplier\n",
    "    plt.figure()\n",
    "    branch_lengths = np.array(branch_lengths)\n",
    "    plt.scatter(branch_lengths[:,0]-branch_lengths[:,1], branch_lengths[:,2])\n",
    "    plt.xlabel(\"stretch\")\n",
    "    plt.ylabel(\"rate multiplier\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "597cc8c8",
   "metadata": {},
   "source": [
    "### Ancestral sequences\n",
    "\n",
    "As mentioned, because model-based phylogenetics uses models to infer phylogenetic trees, we can use those models to infer more than just trees.  \n",
    "One example is that we can infer ancestral sequences from our data, since in building our trees we model the evolution of our sequences down its branches.\n",
    "\n",
    "So finally, let's try inferring ancestral sequences from our covid variants!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19f5d346",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    plt.ion()\n",
    "\n",
    "    # Path to data\n",
    "    base_name = 'covid_variants_aligned'\n",
    "\n",
    "    # Instantiate treetime\n",
    "    myTree = TreeAnc(gtr='Jukes-Cantor', tree=base_name+\".newick\", aln=base_name+'.afa', verbose = 0)\n",
    "\n",
    "    # Infer a GTR model and ML ancestral sequences, either marginally or jointly most likely\n",
    "    myTree.infer_ancestral_sequences(infer_gtr=True, marginal=True)\n",
    "\n",
    "    # Lets examine the properties of a node in the tree after ancestral inference\n",
    "    node = myTree.tree.get_nonterminals()[1]\n",
    "    # Each node now has an inferred sequence\n",
    "    if treetime.version<\"0.7\":\n",
    "        print(\"\\nthe inferred sequences is an array of characters:\", node.sequence)\n",
    "    else:\n",
    "        print(\"\\nthe inferred sequences is an array of characters:\", myTree.sequence(node, as_string=False))\n",
    "\n",
    "    # In addition, each node of the tree now has an mutation object attached\n",
    "    # Note that the mutation numbering starts at 0 rather than 1\n",
    "    print(\"\\nmutations on the branch leading to node %s:\"%node.name, node.mutations)\n",
    "\n",
    "    # We can readily verify these mutations by checking the inferred sequences\n",
    "    if node.mutations:\n",
    "        mut = node.mutations[0]\n",
    "        if treetime.version<\"0.7\":\n",
    "            print(\"\\nmutation %s%d%s corresponds to\"%mut,\n",
    "                  \"parent state: %s, child state %s\\n\\n\"%(node.up.sequence[mut[1]], node.sequence[mut[1]]))\n",
    "        else:\n",
    "            print(\"\\nmutation %s%d%s corresponds to\"%mut,\n",
    "                  \"parent state: %s, child state %s\\n\\n\"%(myTree.sequence(node.up)[mut[1]], myTree.sequence(node)[mut[1]]))\n",
    "\n",
    "\n",
    "    # Plot the tree and label branches by mutation\n",
    "    fig, ax = plt.subplots(1,1,figsize=(18,13))\n",
    "    ax.set_title(\"branches annotated by inferred mutations\", fontsize=18)\n",
    "    ax.set_axis_off()\n",
    "    Phylo.draw(myTree.tree, label_func = lambda x:\"\", axes=ax,\n",
    "        branch_labels=lambda x:\", \".join(map(lambda m: \"%s%d%s\"%m, x.mutations[:3]))\n",
    "                               +('...' if len(x.mutations)>3 else \"\"))\n",
    "\n",
    "    # Finally, we print the inferred GTR model\n",
    "    print(myTree.gtr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8997f4f-2376-47e2-8b82-50727664a2f6",
   "metadata": {},
   "source": [
    "Want to add some more variants to the plot.  See here for a much larger tree:\n",
    "https://www.bv-brc.org/view/VariantLineage/#view_tab=phlyogeny"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BIOM480A5",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "448px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
